### PR TITLE
Facet by Meta

### PR DESCRIPTION
**Pull request by felipeelia**
### Description of the Change
This PR is the following step to #2919, implementing a new type for meta fields.

It still needs tests.

### Changelog Entry

> Added - New block to facet by meta fields

### Credits

Props @felipeelia

### CODE REVIEW

File: includes/classes/Feature/Facets/Types/Meta/Block.php [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+	/**
+	 * Outputs the block preview
+	 *
+	 * @param \WP_REST_Request $request REST request
+	 * @return string
+	 */
+	public function render_block_preview( $request ) {
+		global $wp_query;
+
+		add_filter( 'ep_is_facetable', '__return_true' );
```
Comment (tott) [inline includes/classes/Feature/Facets/Types/Meta/Block.php:140]: would it make sense to provide more context in the filter to be able to make a decision?

File: includes/classes/Feature/Facets/Types/Meta/FacetType.php [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+		 * @return {null|array} Distinct meta values array or `null` to keep default behavior.
+		 */
+		$custom_meta_values = apply_filters( 'ep_facet_meta_custom_meta_values', null, $meta_key );
+		if ( null !== $custom_meta_values ) {
+			return $custom_meta_values;
+		}
+
+		$meta_values = get_transient( self::TRANSIENT_PREFIX . $meta_key );
+		if ( ! $meta_values ) {
+			$meta_values = \ElasticPress\Indexables::factory()->get( 'post' )->get_all_distinct_values( "meta.{$meta_key}.raw" );
```
Comment (tott) [inline includes/classes/Feature/Facets/Types/Meta/FacetType.php]: wondering if we need to truncate those values. there could be huge data in meta fields eventually. maybe truncate out only first 100 chars or so?

**Review by tott — COMMENTED** [review]
Just some minor comments. One thing i'm wondering though - How will things work when people have multiple facet blocks of the same type. I did not see any variation in filters based on the block id or the like. Have you tested this?

File: includes/classes/Feature/Facets/Types/Meta/FacetType.php (hunk 2 above)
Comment (felipeelia) [reply includes/classes/Feature/Facets/Types/Meta/FacetType.php]: As I can see some usefulness in bringing more data in other scenarios, in 2626fb4 I've added the number as a parameter.

File: includes/classes/Feature/Facets/Types/Meta/Block.php (hunk 1 above)
Comment (felipeelia) [reply includes/classes/Feature/Facets/Types/Meta/Block.php:140]: Addressed in f9ae942. I will apply a similar change to #2919 as well, especially for the renderer class part.

**Comment by felipeelia** [comment]
With the change made in 7c1b813, this is the code needed to make Facet by Meta blocks available:

```
add_filter(
	'ep_facet_types',
	function ( $facet_types ) {
		if ( ! isset( $facet_types['meta'] ) && class_exists( '\ElasticPress\Feature\Facets\Types\Meta\FacetType' ) ) {
			$facet_types['meta'] = '\ElasticPress\Feature\Facets\Types\Meta\FacetType';
		}
		return $facet_types;
	}
);
```

**Pull request merged**